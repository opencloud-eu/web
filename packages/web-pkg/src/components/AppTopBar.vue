<template>
  <portal to="app.runtime.header.left">
    <div class="oc-app-top-bar oc-flex">
      <span class="oc-app-top-bar-inner oc-pl-m oc-pr-xs oc-flex oc-flex-middle oc-flex-between">
        <div class="open-file-bar oc-flex">
          <resource-list-item
            v-if="resource"
            id="app-top-bar-resource"
            :is-thumbnail-displayed="false"
            :is-extension-displayed="areFileExtensionsShown"
            :path-prefix="getPathPrefix(resource)"
            :resource="resource"
            :parent-folder-name="getParentFolderName(resource)"
            :parent-folder-link-icon-additional-attributes="
              getParentFolderLinkIconAdditionalAttributes(resource)
            "
            :is-path-displayed="isPathDisplayed"
            :is-resource-clickable="false"
          />
        </div>
        <div class="oc-flex main-actions">
          <template v-if="dropDownMenuSections.length">
            <oc-button
              id="oc-openfile-contextmenu-trigger"
              v-oc-tooltip="contextMenuLabel"
              :aria-label="contextMenuLabel"
              appearance="raw-inverse"
              color-role="chrome"
              class="oc-p-xs"
            >
              <oc-icon name="more-2" />
            </oc-button>
            <oc-drop
              drop-id="oc-openfile-contextmenu"
              mode="click"
              padding-size="small"
              toggle="#oc-openfile-contextmenu-trigger"
              close-on-click
              :title="resource.name"
              @click.stop.prevent
            >
              <context-action-menu
                :menu-sections="dropDownMenuSections"
                :action-options="dropDownActionOptions"
              />
            </oc-drop>
          </template>
          <span v-if="hasAutosave" class="oc-flex oc-flex-middle">
            <oc-icon
              v-oc-tooltip="autoSaveTooltipText"
              :accessible-label="autoSaveTooltipText"
              name="refresh"
              color="white"
              class="ox-p-xs oc-mx-xs"
            />
          </span>
          <template v-if="mainActions.length && resource">
            <context-action-menu
              :menu-sections="[
                {
                  name: 'main-actions',
                  items: mainActions
                    .filter((action) => action.isVisible())
                    .map((action) => {
                      return {
                        ...action,
                        class: 'oc-p-xs app-topbar-action',
                        hideLabel: true
                      }
                    })
                }
              ]"
              :action-options="{
                resources: [resource]
              }"
              appearance="raw-inverse"
              color-role="chrome"
            />
          </template>
          <oc-button
            id="app-top-bar-close"
            appearance="raw-inverse"
            color-role="chrome"
            class="oc-p-xs"
            :aria-label="$gettext('Close')"
            @click="$emit('close')"
          >
            <oc-icon name="close" />
          </oc-button>
        </div>
      </span>
    </div>
  </portal>
</template>

<script lang="ts">
import { computed, defineComponent, PropType, unref } from 'vue'
import ContextActionMenu from './ContextActions/ContextActionMenu.vue'
import { useGettext } from 'vue3-gettext'
import {
  Action,
  FileActionOptions,
  useConfigStore,
  useFolderLink,
  useGetMatchingSpace,
  useResourcesStore
} from '../composables'
import ResourceListItem from './FilesList/ResourceListItem.vue'
import { isPublicSpaceResource, Resource } from '@opencloud-eu/web-client'
import { Duration } from 'luxon'
import { MenuSection } from './ContextActions'

export default defineComponent({
  name: 'AppTopBar',
  components: {
    ContextActionMenu,
    ResourceListItem
  },
  props: {
    dropDownMenuSections: {
      type: Array as PropType<MenuSection[]>,
      default: (): MenuSection[] => []
    },
    dropDownActionOptions: {
      type: Object as PropType<FileActionOptions>,
      default: (): FileActionOptions => ({
        space: null,
        resources: []
      })
    },
    mainActions: {
      type: Array as PropType<Action[]>,
      default: (): Action[] => []
    },
    hasAutoSave: {
      type: Boolean,
      default: true
    },
    isEditor: {
      type: Boolean,
      default: false
    },
    resource: {
      type: Object as PropType<Resource>,
      default: null
    }
  },
  emits: ['close'],
  setup(props) {
    const { $gettext, current: currentLanguage } = useGettext()
    const resourcesStore = useResourcesStore()
    const configStore = useConfigStore()
    const { getMatchingSpace } = useGetMatchingSpace()

    const areFileExtensionsShown = computed(() => resourcesStore.areFileExtensionsShown)
    const contextMenuLabel = computed(() => $gettext('Show context menu'))
    const hasAutosave = computed(
      () => props.isEditor && props.hasAutoSave && configStore.options.editor.autosaveEnabled
    )
    const autoSaveTooltipText = computed(() => {
      const duration = Duration.fromObject(
        { seconds: configStore.options.editor.autosaveInterval },
        { locale: currentLanguage }
      )
      return $gettext(`Autosave (every %{ duration })`, { duration: duration.toHuman() })
    })

    const space = computed(() => getMatchingSpace(props.resource))

    const isPathDisplayed = computed(() => {
      return !isPublicSpaceResource(unref(space))
    })

    return {
      contextMenuLabel,
      areFileExtensionsShown,
      hasAutosave,
      autoSaveTooltipText,
      isPathDisplayed,
      ...useFolderLink()
    }
  }
})
</script>

<style lang="scss">
.oc-app-top-bar {
  align-self: center;
  grid-column: 1 / 4;
  grid-row: secondRow;

  @media (min-width: $oc-breakpoint-small-default) {
    grid-column: 2;
    grid-row: 1;
  }
}

.oc-app-top-bar-inner {
  align-self: center;
  background-color: var(--oc-role-chrome);
  border-radius: 10px;
  border: 1px solid var(--oc-role-on-chrome);
  display: inline-flex;
  gap: 25px;
  height: 40px;
  margin: 10px auto;
  width: 100%;

  @media (min-width: $oc-breakpoint-small-default) {
    flex-basis: 250px;
    margin: 0;
  }

  .oc-resource-indicators {
    .text {
      color: var(--oc-role-on-chrome);
    }
  }
}

.app-topbar-action {
  color: var(--oc-role-on-chrome) !important;

  svg {
    fill: var(--oc-role-on-chrome) !important;
  }

  &:hover:not(:disabled) {
    color: var(--oc-role-on-surface) !important;

    svg {
      fill: var(--oc-role-on-surface) !important;
    }
  }
}

.open-file-bar {
  #app-top-bar-resource {
    max-width: 360px;

    @media (max-width: $oc-breakpoint-medium-default) {
      max-width: 240px;
    }

    @media (min-width: $oc-breakpoint-small-default) {
      widows: initial;
    }

    svg,
    .oc-resource-name span {
      fill: var(--oc-role-on-chrome) !important;
      color: var(--oc-role-on-chrome) !important;
    }
  }

  .oc-resource-icon:hover,
  .oc-resource-name:hover {
    cursor: default;
    text-decoration: none;
  }
}
</style>
