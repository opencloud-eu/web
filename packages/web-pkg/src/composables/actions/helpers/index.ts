export * from './useIsSearchActive'
export * from './useFileActionsDeleteResources'
export * from './useIsFilesAppActive'
export * from './useIsAppActive'
