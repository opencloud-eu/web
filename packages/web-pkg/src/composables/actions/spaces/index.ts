export * from './useSpaceActionsDelete'
export * from './useSpaceActionsDisable'
export * from './useSpaceActionsDuplicate'
export * from './useSpaceActionsEditDescription'
export * from './useSpaceActionsEditQuota'
export * from './useSpaceActionsEditReadmeContent'
export * from './useSpaceActionsRename'
export * from './useSpaceActionsRestore'
export * from './useSpaceActionsShowMembers'
export * from './useSpaceActionsNavigateToTrash'
export * from './useSpaceActionsSetIcon'
export * from './useSpaceActionsDeleteImage'
