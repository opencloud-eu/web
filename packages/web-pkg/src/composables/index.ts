export * from './ability'
export * from './actions'
export * from './appDefaults'
export * from './appProviderService'
export * from './archiverService'
export * from './authContext'
export * from './breadcrumbs'
export * from './clientService'
export * from './clipboard'
export * from './download'
export * from './driveResolver'
export * from './embedMode'
export * from './eventBus'
export * from './fileListHeaderPosition'
export * from './filesList'
export * from './folderLink'
export * from './isTopBarSticky'
export * from './keyboardActions'
export * from './links'
export * from './loadingService'
export * from './localStorage'
export * from './pagination'
export * from './passwordPolicyService'
export * from './piniaStores'
export * from './portalTarget'
export * from './previewService'
export * from './requestHeaders'
export * from './resources'
export * from './router'
export * from './scrollTo'
export * from './search'
export * from './selection'
export * from './service'
export * from './shares'
export * from './sideBar'
export * from './sort'
export * from './spaces'
export * from './upload'
export * from './viewMode'
export * from './webWorkers'
export * from './avatars'
