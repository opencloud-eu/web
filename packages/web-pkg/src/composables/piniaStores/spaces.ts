import { defineStore } from 'pinia'
import { computed, ref, unref } from 'vue'
import {
  buildShareSpaceResource,
  isMountPointSpaceResource,
  SpaceResource
} from '@opencloud-eu/web-client'
import { Graph } from '@opencloud-eu/web-client/graph'
import {
  buildSpace,
  extractStorageId,
  isPersonalSpaceResource,
  isProjectSpaceResource
} from '@opencloud-eu/web-client'
import type {
  CollaboratorShare,
  MountPointSpaceResource,
  ShareRole
} from '@opencloud-eu/web-client'
import { useUserStore } from './user'
import { ConfigStore, useConfigStore } from './config'
import { useSharesStore } from './shares'

// sort space members with higher permissions (managers) at the top
export const sortSpaceMembers = (shares: CollaboratorShare[]) => {
  return shares.sort((a, b) => b.permissions.length - a.permissions.length)
}

export const getSpacesByType = async ({
  graphClient,
  driveType,
  configStore,
  graphRoles,
  signal
}: {
  graphClient: Graph
  driveType: string
  configStore: ConfigStore
  graphRoles: Record<string, ShareRole>
  signal?: AbortSignal
}) => {
  const mountpoints = await graphClient.drives.listMyDrives(
    graphRoles,
    {
      orderBy: 'name asc',
      filter: `driveType eq ${driveType}`
    },
    { signal }
  )
  if (!mountpoints.length) {
    return []
  }

  if (driveType !== 'mountpoint' || !configStore.options.routing?.fullShareOwnerPaths) {
    return mountpoints
  }

  const rootSpaceDriveAliasMapping: Record<string, string> = {}
  mountpoints.forEach((space) => {
    const { rootId, driveAlias } = space.root.remoteItem
    rootSpaceDriveAliasMapping[rootId] = driveAlias
  })

  const rootSpaces = Object.entries(rootSpaceDriveAliasMapping).map(([id, driveAlias]) =>
    // FIXME: create proper buildRootSpace (or whatever function)
    buildSpace(
      {
        id: extractStorageId(id),
        name: driveAlias, // FIXME: set a proper name
        driveType: driveAlias.split('/')[0], // FIXME: can we retrieve this from api?
        driveAlias,
        path: '/',
        serverUrl: configStore.serverUrl
      },
      graphRoles
    )
  )

  return [...mountpoints, ...rootSpaces]
}

export const useSpacesStore = defineStore('spaces', () => {
  const userStore = useUserStore()
  const configStore = useConfigStore()
  const sharesStore = useSharesStore()

  const spaces = ref<SpaceResource[]>([])
  const currentSpace = ref<SpaceResource>()
  const spacesInitialized = ref(false)
  const mountPointsInitialized = ref(false)
  const spacesLoading = ref(false)
  const defaultSpaceImageBlobURL = ref<string>(null)
  const imagesLoading = ref<string[]>([])
  const readmesLoading = ref<string[]>([])

  const personalSpace = computed(() => {
    return unref(spaces).find((s) => isPersonalSpaceResource(s) && s.isOwner(userStore.user))
  })

  const setSpacesInitialized = (value: boolean) => {
    spacesInitialized.value = value
  }

  const setMountPointsInitialized = (value: boolean) => {
    mountPointsInitialized.value = value
  }

  const setSpacesLoading = (value: boolean) => {
    spacesLoading.value = value
  }

  const setCurrentSpace = (space: SpaceResource) => {
    currentSpace.value = space
  }

  const setDefaultSpaceImageBlobURL = (blobURL: string) => {
    defaultSpaceImageBlobURL.value = blobURL
  }

  const getSpaceMembers = (space: SpaceResource) => {
    // only project spaces have members
    if (!isProjectSpaceResource(space)) {
      return []
    }
    const members = sharesStore.collaboratorShares.filter((c) => c.resourceId === space.id)
    return sortSpaceMembers(members)
  }

  const addSpaces = (s: SpaceResource[]) => {
    unref(spaces).push(...s)
  }

  const removeSpace = (space: SpaceResource) => {
    spaces.value = unref(spaces).filter(({ id }) => id !== space.id)
  }

  const getSpace = (id: string) => {
    return unref(spaces).find((s) => id == s.id)
  }

  const getMountPointForSpace = async ({
    graphClient,
    space,
    signal
  }: {
    graphClient: Graph
    space: SpaceResource
    signal?: AbortSignal
  }): Promise<MountPointSpaceResource> => {
    await loadMountPoints({ graphClient, signal })

    // even if the resource has been shared via multiple permissions (e.g. directly via user and a group)
    // we only care about one matching mount point since the remote item contains all permissions
    return unref(spaces).find(
      (s) => isMountPointSpaceResource(s) && s.root?.remoteItem?.id === space.id
    )
  }

  const createShareSpace = ({
    driveAliasPrefix,
    id,
    shareName
  }: {
    driveAliasPrefix: 'share' | 'ocm-share'
    id: string
    shareName: string
  }) => {
    const space = buildShareSpaceResource({
      driveAliasPrefix,
      id,
      shareName,
      serverUrl: configStore.serverUrl
    })
    addSpaces([space])
    return space
  }

  const upsertSpace = (space: SpaceResource) => {
    const existingSpace = unref(spaces).find(({ id }) => id === space.id)
    if (existingSpace) {
      Object.assign(existingSpace, space)
      return
    }
    addSpaces([space])
  }

  const updateSpaceField = <T extends SpaceResource>({
    id,
    field,
    value
  }: {
    id: T['id']
    field: keyof T
    value: T[keyof T]
  }) => {
    const space = unref(spaces).find((space) => id === space.id) as T
    if (space) {
      space[field] = value
    }
  }

  const loadSpaces = async ({ graphClient }: { graphClient: Graph }) => {
    spacesLoading.value = true
    try {
      /**
       * FIXME: this is bad for two reasons:
       * 1. fetching by specific drive type is bad because if more drive types are being added it needs additional code.
       *    as soon as the backend allows to filter by `driveType neq virtual` we want to use that here.
       * 2. fetching the mountpoint drives only on first access is kind of error prone, because mount points are
       *    trying to be accessed in multiple code locations. all of them need to check now if mountpoints need to be
       *    fetched first. but at the moment fetching mountpoints is kind of expensive, so we need to accept that for now.
       */
      const [personalSpaces, projectSpaces] = await Promise.all([
        getSpacesByType({
          graphClient,
          driveType: 'personal',
          configStore,
          graphRoles: sharesStore.graphRoles
        }),
        getSpacesByType({
          graphClient,
          driveType: 'project',
          configStore,
          graphRoles: sharesStore.graphRoles
        })
      ])

      addSpaces([...personalSpaces, ...projectSpaces])
      spacesInitialized.value = true
    } finally {
      spacesLoading.value = false
    }
  }

  const loadMountPoints = async ({
    graphClient,
    signal
  }: {
    graphClient: Graph
    signal?: AbortSignal
  }) => {
    // fetching mount points is particularly expensive, so we do that only on first access.
    if (unref(mountPointsInitialized)) {
      return
    }
    try {
      const mountPointSpaces = await getSpacesByType({
        graphClient,
        driveType: 'mountpoint',
        configStore,
        graphRoles: sharesStore.graphRoles,
        signal
      })
      addSpaces(mountPointSpaces)
    } finally {
      mountPointsInitialized.value = true
    }
  }

  const reloadProjectSpaces = async ({
    graphClient,
    signal
  }: {
    graphClient: Graph
    signal?: AbortSignal
  }) => {
    const projectSpaces = await getSpacesByType({
      graphClient,
      driveType: 'project',
      configStore,
      graphRoles: sharesStore.graphRoles,
      signal
    })
    spaces.value = unref(spaces).filter((s) => !isProjectSpaceResource(s))
    addSpaces(projectSpaces)
  }

  const addToImagesLoading = (id: string) => {
    if (!unref(imagesLoading).includes(id)) {
      unref(imagesLoading).push(id)
    }
  }

  const removeFromImagesLoading = (id: string) => {
    imagesLoading.value = unref(imagesLoading).filter((i) => i !== id)
  }

  const purgeImagesLoading = () => {
    imagesLoading.value = []
  }

  const addToReadmesLoading = (id: string) => {
    if (!unref(readmesLoading).includes(id)) {
      unref(readmesLoading).push(id)
    }
  }

  const removeFromReadmesLoading = (id: string) => {
    readmesLoading.value = unref(readmesLoading).filter((i) => i !== id)
  }

  const purgeReadmesLoading = () => {
    readmesLoading.value = []
  }

  return {
    spaces,
    spacesInitialized,
    mountPointsInitialized,
    spacesLoading,
    currentSpace,
    personalSpace,
    defaultSpaceImageBlobURL,
    imagesLoading,
    readmesLoading,

    getSpace,
    createShareSpace,
    setSpacesInitialized,
    setMountPointsInitialized,
    setSpacesLoading,
    setCurrentSpace,
    setDefaultSpaceImageBlobURL,
    getSpaceMembers,
    getMountPointForSpace,

    addSpaces,
    removeSpace,
    upsertSpace,
    updateSpaceField,
    loadSpaces,
    loadMountPoints,
    reloadProjectSpaces,

    addToImagesLoading,
    removeFromImagesLoading,
    purgeImagesLoading,
    addToReadmesLoading,
    removeFromReadmesLoading,
    purgeReadmesLoading
  }
})

export type SpacesStore = ReturnType<typeof useSpacesStore>
