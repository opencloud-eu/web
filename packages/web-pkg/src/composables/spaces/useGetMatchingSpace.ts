import { useRouteParam } from '../router'
import { Resource, SpaceResource } from '@opencloud-eu/web-client'
import {
  MountPointSpaceResource,
  extractStorageId,
  isMountPointSpaceResource,
  isProjectSpaceResource,
  ShareTypes,
  OCM_PROVIDER_ID,
  isShareResource
} from '@opencloud-eu/web-client'
import { computed, Ref, unref } from 'vue'
import { basename } from 'path'
import { useSpacesStore, useUserStore, useConfigStore } from '../piniaStores'

type GetMatchingSpaceOptions = {
  space?: Ref<SpaceResource>
}

export const useGetMatchingSpace = (options?: GetMatchingSpaceOptions) => {
  const userStore = useUserStore()
  const spacesStore = useSpacesStore()
  const configStore = useConfigStore()
  const spaces = computed(() => spacesStore.spaces)
  const driveAliasAndItem = useRouteParam('driveAliasAndItem')

  const getInternalSpace = (storageId: string): SpaceResource => {
    return unref(options?.space) || unref(spaces).find((space) => space.id === storageId)
  }

  const getMatchingSpace = (resource: Resource): SpaceResource => {
    let storageId = resource.storageId

    if (
      unref(driveAliasAndItem)?.startsWith('public/') ||
      unref(driveAliasAndItem)?.startsWith('ocm/')
    ) {
      storageId = unref(driveAliasAndItem).split('/')[1]
    }

    const space = getInternalSpace(storageId)

    if (space && !isMountPointSpaceResource(space)) {
      return space
    }

    const driveAliasPrefix =
      (isShareResource(resource) && resource.shareTypes.includes(ShareTypes.remote.value)) ||
      resource?.id?.toString().startsWith(OCM_PROVIDER_ID)
        ? 'ocm-share'
        : 'share'

    let shareName: string
    if (resource.remoteItemPath) {
      shareName = basename(resource.remoteItemPath)
    } else if (
      unref(driveAliasAndItem)?.startsWith('share/') ||
      unref(driveAliasAndItem)?.startsWith('ocm-share/')
    ) {
      shareName = unref(driveAliasAndItem).split('/')[1]
    } else {
      shareName = resource.name
    }

    return (
      spacesStore.getSpace(resource.remoteItemId) ||
      spacesStore.createShareSpace({
        driveAliasPrefix,
        id: resource.remoteItemId,
        shareName
      })
    )
  }

  const getMatchingMountPoints = (space: SpaceResource): MountPointSpaceResource[] =>
    unref(spaces).filter(
      (s) => isMountPointSpaceResource(s) && extractStorageId(s.root.remoteItem.rootId) === space.id
    )

  const isPersonalSpaceRoot = (resource: Resource) => {
    return (
      resource?.storageId &&
      resource?.storageId === spacesStore.personalSpace?.storageId &&
      resource?.path === '/'
    )
  }

  const isResourceAccessible = ({ space, path }: { space: SpaceResource; path: string }) => {
    if (!configStore.options.routing.fullShareOwnerPaths) {
      return true
    }

    const projectSpace = unref(spaces).some((s) => isProjectSpaceResource(s) && s.id === space.id)
    const fullyAccessibleSpace = space.isOwner(userStore.user) || projectSpace

    return (
      fullyAccessibleSpace ||
      getMatchingMountPoints(space).some((m) => path.startsWith(m.root.remoteItem.path))
    )
  }

  return {
    getInternalSpace,
    getMatchingSpace,
    isPersonalSpaceRoot,
    isResourceAccessible
  }
}
