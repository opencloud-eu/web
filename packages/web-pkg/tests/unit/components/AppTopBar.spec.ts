import { mock } from 'vitest-mock-extended'
import {
  RouteLocation,
  defaultComponentMocks,
  defaultPlugins,
  shallowMount,
  useGetMatchingSpaceMock
} from '@opencloud-eu/web-test-helpers'
import { Resource, SpaceResource } from '@opencloud-eu/web-client'
import AppTopBar from '../../../src/components/AppTopBar.vue'
import { Action } from '../../../src/composables/actions'
import { useGetMatchingSpace } from '../../../src/composables/spaces/useGetMatchingSpace'

vi.mock('../../../src/composables/spaces/useGetMatchingSpace')

describe('AppTopBar', () => {
  describe('if no resource is present', () => {
    it('renders only a close button', () => {
      const { wrapper } = getWrapper(mock<Resource>({ path: '/test.txt', remoteItemPath: '/test' }))
      expect(wrapper.html()).toMatchSnapshot()
    })
  })
  describe('if a resource is present', () => {
    it('renders a resource and no actions (if none given) and a close button', () => {
      const { wrapper } = getWrapper(mock<Resource>({ path: '/test.txt', remoteItemPath: '/test' }))
      expect(wrapper.html()).toMatchSnapshot()
    })

    it('renders a resource and mainActions (if given) and a close button', () => {
      const { wrapper } = getWrapper(
        mock<Resource>({ path: '/test.txt', remoteItemPath: '/test' }),
        [],
        [mock<Action>()]
      )
      expect(wrapper.html()).toMatchSnapshot()
    })

    it('renders a resource and dropdownActions (if given) and a close button', () => {
      const { wrapper } = getWrapper(
        mock<Resource>({ path: '/test.txt', remoteItemPath: '/test' }),
        [mock<Action>()],
        []
      )
      expect(wrapper.html()).toMatchSnapshot()
    })
    it('renders a resource and dropdownActions as well as mainActions (if both are passed) and a close button', () => {
      const { wrapper } = getWrapper(
        mock<Resource>({ path: '/test.txt', remoteItemPath: '/test' }),
        [mock<Action>()],
        [mock<Action>()]
      )
      expect(wrapper.html()).toMatchSnapshot()
    })
    it('renders a resource without file extension if areFileExtensionsShown is set to false', () => {
      const { wrapper } = getWrapper(
        mock<Resource>({ path: '/test.txt', remoteItemPath: '/test' }),
        [mock<Action>()],
        [mock<Action>()],
        false
      )

      expect(wrapper.html()).toMatchSnapshot()
    })
  })
})

function getWrapper(
  resource: Resource = null,
  dropDownActions: Action[] = [],
  mainActions: Action[] = [],
  areFileExtensionsShown = true
) {
  const mocks = defaultComponentMocks({
    currentRoute: mock<RouteLocation>({ name: 'admin-settings-general' })
  })

  vi.mocked(useGetMatchingSpace).mockImplementation(() =>
    useGetMatchingSpaceMock({
      getInternalSpace: () => mock<SpaceResource>(),
      isResourceAccessible: () => true
    })
  )

  return {
    wrapper: shallowMount(AppTopBar, {
      props: {
        dropDownActions,
        mainActions,
        resource
      },
      global: {
        plugins: [
          ...defaultPlugins({ piniaOptions: { resourcesStore: { areFileExtensionsShown } } })
        ],
        mocks,
        provide: mocks
      }
    })
  }
}
