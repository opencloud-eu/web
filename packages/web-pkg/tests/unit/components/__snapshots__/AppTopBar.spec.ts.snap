// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AppTopBar > if a resource is present > renders a resource and dropdownActions (if given) and a close button 1`] = `
"<portal to="app.runtime.header.left" dropdownactions="undefined">
  <div class="oc-app-top-bar oc-flex"><span class="oc-app-top-bar-inner oc-pl-m oc-pr-xs oc-flex oc-flex-middle oc-flex-between"><div class="open-file-bar oc-flex"><resource-list-item-stub resource="undefined" pathprefix="[Function]" ispathdisplayed="true" parentfoldername="Personal" parentfolderlinkiconadditionalattributes="[object Object]" isextensiondisplayed="true" isthumbnaildisplayed="false" isicondisplayed="true" isresourceclickable="false" id="app-top-bar-resource"></resource-list-item-stub></div> <div class="oc-flex main-actions"><!--v-if--> <!--v-if--> <!--v-if--> <oc-button-stub arialabel="Close" appearance="raw-inverse" colorrole="chrome" disabled="false" gapsize="medium" justifycontent="center" showspinner="false" size="medium" submit="button" type="button" nohover="false" id="app-top-bar-close" class="oc-p-xs"></oc-button-stub></div></span></div>
</portal>"
`;

exports[`AppTopBar > if a resource is present > renders a resource and dropdownActions as well as mainActions (if both are passed) and a close button 1`] = `
"<portal to="app.runtime.header.left" dropdownactions="undefined">
  <div class="oc-app-top-bar oc-flex"><span class="oc-app-top-bar-inner oc-pl-m oc-pr-xs oc-flex oc-flex-middle oc-flex-between"><div class="open-file-bar oc-flex"><resource-list-item-stub resource="undefined" pathprefix="[Function]" ispathdisplayed="true" parentfoldername="Personal" parentfolderlinkiconadditionalattributes="[object Object]" isextensiondisplayed="true" isthumbnaildisplayed="false" isicondisplayed="true" isresourceclickable="false" id="app-top-bar-resource"></resource-list-item-stub></div> <div class="oc-flex main-actions"><!--v-if--> <!--v-if--> <context-action-menu-stub menusections="[object Object]" actionoptions="[object Object]" appearance="raw-inverse" color-role="chrome"></context-action-menu-stub> <oc-button-stub arialabel="Close" appearance="raw-inverse" colorrole="chrome" disabled="false" gapsize="medium" justifycontent="center" showspinner="false" size="medium" submit="button" type="button" nohover="false" id="app-top-bar-close" class="oc-p-xs"></oc-button-stub></div></span></div>
</portal>"
`;

exports[`AppTopBar > if a resource is present > renders a resource and mainActions (if given) and a close button 1`] = `
"<portal to="app.runtime.header.left" dropdownactions="">
  <div class="oc-app-top-bar oc-flex"><span class="oc-app-top-bar-inner oc-pl-m oc-pr-xs oc-flex oc-flex-middle oc-flex-between"><div class="open-file-bar oc-flex"><resource-list-item-stub resource="undefined" pathprefix="[Function]" ispathdisplayed="true" parentfoldername="Personal" parentfolderlinkiconadditionalattributes="[object Object]" isextensiondisplayed="true" isthumbnaildisplayed="false" isicondisplayed="true" isresourceclickable="false" id="app-top-bar-resource"></resource-list-item-stub></div> <div class="oc-flex main-actions"><!--v-if--> <!--v-if--> <context-action-menu-stub menusections="[object Object]" actionoptions="[object Object]" appearance="raw-inverse" color-role="chrome"></context-action-menu-stub> <oc-button-stub arialabel="Close" appearance="raw-inverse" colorrole="chrome" disabled="false" gapsize="medium" justifycontent="center" showspinner="false" size="medium" submit="button" type="button" nohover="false" id="app-top-bar-close" class="oc-p-xs"></oc-button-stub></div></span></div>
</portal>"
`;

exports[`AppTopBar > if a resource is present > renders a resource and no actions (if none given) and a close button 1`] = `
"<portal to="app.runtime.header.left" dropdownactions="">
  <div class="oc-app-top-bar oc-flex"><span class="oc-app-top-bar-inner oc-pl-m oc-pr-xs oc-flex oc-flex-middle oc-flex-between"><div class="open-file-bar oc-flex"><resource-list-item-stub resource="undefined" pathprefix="[Function]" ispathdisplayed="true" parentfoldername="Personal" parentfolderlinkiconadditionalattributes="[object Object]" isextensiondisplayed="true" isthumbnaildisplayed="false" isicondisplayed="true" isresourceclickable="false" id="app-top-bar-resource"></resource-list-item-stub></div> <div class="oc-flex main-actions"><!--v-if--> <!--v-if--> <!--v-if--> <oc-button-stub arialabel="Close" appearance="raw-inverse" colorrole="chrome" disabled="false" gapsize="medium" justifycontent="center" showspinner="false" size="medium" submit="button" type="button" nohover="false" id="app-top-bar-close" class="oc-p-xs"></oc-button-stub></div></span></div>
</portal>"
`;

exports[`AppTopBar > if a resource is present > renders a resource without file extension if areFileExtensionsShown is set to false 1`] = `
"<portal to="app.runtime.header.left" dropdownactions="undefined">
  <div class="oc-app-top-bar oc-flex"><span class="oc-app-top-bar-inner oc-pl-m oc-pr-xs oc-flex oc-flex-middle oc-flex-between"><div class="open-file-bar oc-flex"><resource-list-item-stub resource="undefined" pathprefix="[Function]" ispathdisplayed="true" parentfoldername="Personal" parentfolderlinkiconadditionalattributes="[object Object]" isextensiondisplayed="false" isthumbnaildisplayed="false" isicondisplayed="true" isresourceclickable="false" id="app-top-bar-resource"></resource-list-item-stub></div> <div class="oc-flex main-actions"><!--v-if--> <!--v-if--> <context-action-menu-stub menusections="[object Object]" actionoptions="[object Object]" appearance="raw-inverse" color-role="chrome"></context-action-menu-stub> <oc-button-stub arialabel="Close" appearance="raw-inverse" colorrole="chrome" disabled="false" gapsize="medium" justifycontent="center" showspinner="false" size="medium" submit="button" type="button" nohover="false" id="app-top-bar-close" class="oc-p-xs"></oc-button-stub></div></span></div>
</portal>"
`;

exports[`AppTopBar > if no resource is present > renders only a close button 1`] = `
"<portal to="app.runtime.header.left" dropdownactions="">
  <div class="oc-app-top-bar oc-flex"><span class="oc-app-top-bar-inner oc-pl-m oc-pr-xs oc-flex oc-flex-middle oc-flex-between"><div class="open-file-bar oc-flex"><resource-list-item-stub resource="undefined" pathprefix="[Function]" ispathdisplayed="true" parentfoldername="Personal" parentfolderlinkiconadditionalattributes="[object Object]" isextensiondisplayed="true" isthumbnaildisplayed="false" isicondisplayed="true" isresourceclickable="false" id="app-top-bar-resource"></resource-list-item-stub></div> <div class="oc-flex main-actions"><!--v-if--> <!--v-if--> <!--v-if--> <oc-button-stub arialabel="Close" appearance="raw-inverse" colorrole="chrome" disabled="false" gapsize="medium" justifycontent="center" showspinner="false" size="medium" submit="button" type="button" nohover="false" id="app-top-bar-close" class="oc-p-xs"></oc-button-stub></div></span></div>
</portal>"
`;
